### DISCUSSION TITLE
Most efficient pattern, context manager or commit-as-you-go

### DISCUSSION THREAD

**Discussion by OddBlueDog** [sqlalchemy/sqlalchemy#6797] [Usage Questions] [answered]
What do you think the best way to do the below in SQLalchemy 1.4 / 2.0

Is it a bit resource wasteful to do a session.begin() or commit() for just the select statement? ` result = await session.execute(select(user).filter(user.id == id))`

**1**
```
async with async_session() as session:
    result = await session.execute(select(user).filter(user.id == id))
    # have to commit here to close the transaction, otherwise session.begin() errors that a transaction is begun already on this sesion.
    session.commit()
    ...
    async with session.begin():
        session.add(some_new_user)
```
       
**2**
```
async with async_session() as session:
    async with session.begin():
        result = await session.execute(select(user).filter(user.id == id))
    ...
    async with session.begin():
        session.add(some_new_user)
```

**3**
```
async with async_session() as session:
    result = await session.execute(select(user).filter(user.id == id))
     ...
    session.add(some_new_user)
    session.commit()
```

**Comment by zzzeek** [comment, marked as answer]
> What do you think the best way to do the below in SQLalchemy 1.4 / 2.0
> 
> Is it a bit resource wasteful to do a session.begin() or commit() for just the select statement? ` 

not at all, the DBAPI interaction is identical

> 
> ```
> async with async_session() as session:
>     result = await session.execute(select(user).filter(user.id == id))
>     # have to commit here to close the transaction, otherwise session.begin() errors that a transaction is begun already on this sesion.
>     session.commit()
>     ...
>     async with session.begin():
>         session.add(some_new_user)

for the above I'd just use commit-as-you-go since you are already in a context manager block.  Python DBAPI does not operate without a transaction in play.



> ```
> 
> **2**
> 
> ```
> async with async_session() as session:
>     async with session.begin():
>         result = await session.execute(select(user).filter(user.id == id))
>     ...
>     async with session.begin():
>         session.add(some_new_user)
> ```
> 
> **3**
> 
> ```
> async with async_session() as session:
>     result = await session.execute(select(user).filter(user.id == id))
>      ...
>     session.add(some_new_user)
>     session.commit()
> ```

number 3 looks most succinct, i'd stick with that as it is the most common

**Comment by CaselIT** [comment]
another alternative is to just do the select inside the begin block:

```
async with async_session() as session:
    async with session.begin():
        result = await session.execute(select(user).filter(user.id == id))
        ...
        session.add(some_new_user)
```
